### PR TITLE
User / UserProperty code in `ndb`, for discussion.

### PR DESCRIPTION
**Pull request by dhermes**
I just wanted to get this out for discussion. I'm not sure what to do about the absence of the `google.appengine.api.users.User` type, so wanted to chat about it before I waste too much time writing tests.

### CODE REVIEW

**Comment by chrisrossi** [comment]
I think, in general, if the sole use of something is to provide an integration point with something that is no longer there, then we just leave it out. There will be lots of applications that have a notion of users, but we can't know anything, *a priori*, about the notion of user used in the application layer, down here at the data storage layer. This seems entirely skippable to me. 

But, get buy in from a Googler. I'm no authority here.

**Comment by dhermes** [comment]
The original API had a [`PropertyValue`][3] that could store a [`uservalue`][4] (also see the original [`User`][1] protobuf and the associated [fields][2]). However the new API uses [`Value`][5] which does not have a notion of user.

I need to do some experimenting to determine what comes out the other end when a `UserProperty` is used. I recall that storing a user with email only, when persisted and retrieved the datastore would come back with the user ID as well.

[1]: https://github.com/dhermes/app-engine-standard/blob/19187b658bd11bac1edb0f203c22a14444fc21ff/google/appengine/datastore/entity_pb.py#L1985
[2]: https://github.com/dhermes/app-engine-standard/blob/19187b658bd11bac1edb0f203c22a14444fc21ff/google/appengine/datastore/entity_pb.py#L2265-L2271
[3]: https://github.com/dhermes/app-engine-standard/blob/19187b658bd11bac1edb0f203c22a14444fc21ff/google/appengine/datastore/entity_pb.py#L760
[4]: https://github.com/dhermes/app-engine-standard/blob/19187b658bd11bac1edb0f203c22a14444fc21ff/google/appengine/datastore/entity_pb.py#L1069-L1076
[5]: https://github.com/googleapis/googleapis/blob/e26f9c7cdefd2118823d4af5902471507eb7b3e4/google/datastore/v1/entity.proto#L124

**Comment by dhermes** [comment]
Some GAE standard research. I executed the following snippet in a live app:

```python
import pprint

from google.appengine.api import users
from google.appengine.ext import ndb


u = users.User(email="daniel.j.hermes@gmail.com")
print("u = {!r}".format(u))
print("u.__dict__ =")
pprint.pprint(u.__dict__)


class Simple(ndb.Model):
    u = ndb.UserProperty()


s = Simple(u=u)
print("s = {}".format(s))
s.put()
print("s = {}".format(s))

print("=" * 40)
new_s = s.key.get(use_cache=False, use_memcache=False)
print("new_s = {}".format(new_s))
new_u = new_s.u
print("new_u = {!r}".format(new_u))
print("new_u.__dict__ =")
pprint.pprint(new_u.__dict__)
```

and the output was:

```
u = users.User(email='daniel.j.hermes@gmail.com')
u.__dict__ =
{'_User__auth_domain': 'gmail.com',
 '_User__email': 'daniel.j.hermes@gmail.com',
 '_User__federated_identity': None,
 '_User__federated_provider': None,
 '_User__user_id': None}
s = Simple(u=users.User(email='daniel.j.hermes@gmail.com'))
s = Simple(key=Key('Simple', 5651019216191488),
           u=users.User(email='daniel.j.hermes@gmail.com'))
========================================
new_s = Simple(key=Key('Simple', 5651019216191488),
               u=users.User(email='daniel.j.hermes@gmail.com',_user_id='...redacted...'))
new_u = users.User(email='daniel.j.hermes@gmail.com',_user_id='...redacted...')
new_u.__dict__ =
{'_User__auth_domain': u'gmail.com',
 '_User__email': u'daniel.j.hermes@gmail.com',
 '_User__federated_identity': None,
 '_User__federated_provider': None,
 '_User__user_id': u'...redacted...'}
```

Then querying for this with the new API I got:

```python
>>> import google.cloud.datastore
>>> client = google.cloud.datastore.Client()
>>> key = client.key('Simple', 5651019216191488)
>>> entity = client.get(key)
>>> entity
<Entity('Simple', 5651019216191488) {'u': <Entity {'email': 'daniel.j.hermes@gmail.com', 'user_id': '...redacted...', 'auth_domain': 'gmail.com'}>}>
>>> entity._meanings
{'u': (20, <Entity {'email': 'daniel.j.hermes@gmail.com', ...}>)}
```

Surprising `20` does not show up on the [list of meanings][1], but I do know that `USER_PROTO = 20` is a possible meaning as well.

[1]: https://github.com/dhermes/app-engine-standard/blob/19187b658bd11bac1edb0f203c22a14444fc21ff/google/appengine/datastore/entity_pb.py#L1147-L1167

**Comment by dhermes** [comment]
I've also confirmed that it's possible to insert a `User` property from the new API:

```python
>>> import google.cloud.datastore
>>> client = google.cloud.datastore.Client()
>>> key = client.key('Simple')
>>> e = google.cloud.datastore.Entity(key=key)
>>> e['u'] = google.cloud.datastore.Entity(
...     exclude_from_indexes=('auth_domain', 'email', 'user_id')
... )
>>> e['u']['auth_domain'] = 'gmail.com'
>>> e['u']['email'] = 'daniel.j.hermes@gmail.com'
>>> e['u']['user_id'] = '...redacted...'
>>> e._meanings['u'] = (20, e['u'])
>>> client.put(e)
>>> e.key
<Key('Simple', 5720914775769088), project=precise-truck-742>
```

However the API requires that the `User` sub-properties not be indexed:

```python
>>> e['u'].exclude_from_indexes = set()
>>> e.key = client.key('Simple')
>>> client.put(e)
Traceback (most recent call last):
...
google.api_core.exceptions.InvalidArgument: 400 The user entity property "user_id" is indexed.
```

IOW, the backend knows that `meaning=20` corresponds to a user.

**Comment by dhermes** [comment]
Some more sleuthing.

## `user_id` Round-trip

When using only the new API outside of GAE standard, the `user_id` will not be populated in the round-trip

```python
>>> import google.cloud.datastore
>>> client = google.cloud.datastore.Client()
>>> key = client.key('Simple')
>>> e = google.cloud.datastore.Entity(key=key)
>>> e['u'] = google.cloud.datastore.Entity(
...     exclude_from_indexes=('auth_domain', 'email')
... )
>>> e['u']['auth_domain'] = 'gmail.com'
>>> e['u']['email'] = 'daniel.j.hermes@gmail.com'
>>> e._meanings['u'] = (20, e['u'])
>>> client.put(e)
>>> e
<Entity('Simple', 5661714691391488) {'u': <Entity {'auth_domain': 'gmail.com', 'email': 'daniel.j.hermes@gmail.com'}>}>
```

## Required Fields

We already saw above that `user_id` is not required. We see also that `email` is required

```python
>>> e.key = client.key('Simple')
>>> e['u'].exclude_from_indexes = set(['auth_domain', 'user_id'])
>>> e['u'].clear()
>>> e['u'].update(auth_domain='gmail.com', user_id='...redacted...')
>>> client.put(e)
Traceback (most recent call last):
  ...
InvalidArgument: 400 The user entity is missing required property "email".
```

and that `auth_domain` is required:

```python
>>> e['u'].exclude_from_indexes = set(['email', 'user_id'])
>>> e['u'].clear()
>>> e['u'].update(email='daniel.j.hermes@gmail.com', user_id='...redacted...')
>>> client.put(e)
Traceback (most recent call last):
  ...
InvalidArgument: 400 The user entity is missing required property "auth_domain".
```

## Restricted Fields

Trying to add **other** fields is also not allowed when `meaning=20`

```python
>>> e['u'].exclude_from_indexes = set(['auth_domain', 'email'])
>>> e['u'].clear()
>>> e['u'].update(auth_domain='gmail.com', email='daniel.j.hermes@gmail.com', other=10)
>>> client.put(e)
Traceback (most recent call last):
  ...
InvalidArgument: 400 The user entity property "other" is not allowed.
```

## Indexed Fields

As we saw above `user_id` must be indexed, but we can see that the other fields must be as well:

```python
>>> e['u'].exclude_from_indexes = set(['user_id'])
>>> e['u'].clear()
>>> e['u'].update(auth_domain='gmail.com', email='daniel.j.hermes@gmail.com', user_id='...redacted...')
>>> client.put(e)
Traceback (most recent call last):
  ...
InvalidArgument: 400 The user entity property "email" is indexed.
```

and

```python
>>> e['u'].exclude_from_indexes = set(['email', 'user_id'])
>>> client.put(e)
Traceback (most recent call last):
  ...
InvalidArgument: 400 The user entity property "auth_domain" is indexed.
```

**Comment by chrisrossi** [comment]
Ok. Ignore my first comment. I forgot that the Google Cloud is still going to have a notion of users.

**Comment by dhermes** [comment]
I also determined that `nickname` cannot be set in the new API, but `federated_identity` and `federated_provider` can (and they must not be indexed). I set `federated_identity="A"` and `federated_provider="B"` on an entity with `google-cloud-datastore` but only the identity came back on GAE. Running:

```python
key = ndb.Key(Simple, 5760524373458944)
new_s = key.get(use_cache=False, use_memcache=False)
print("new_s = {}".format(new_s))
new_u = new_s.u
print("new_u = {!r}".format(new_u))
print("new_u.__dict__ =")
pprint.pprint(new_u.__dict__)
```

resulted in

```
new_s = Simple(key=Key('Simple', 5760524373458944), u=users.User(email='daniel.j.hermes@gmail.com',federated_identity='A',_user_id='...redacted...'))
new_u = users.User(email='daniel.j.hermes@gmail.com',federated_identity='A',_user_id='...redacted...')
new_u.__dict__ =
{'_User__auth_domain': u'gmail.com',
 '_User__email': u'daniel.j.hermes@gmail.com',
 '_User__federated_identity': u'A',
 '_User__federated_provider': None,
 '_User__user_id': u'...redacted...'}
```

though I can confirm all fields come back from the new API:

```python
>>> client.get(key)
<Entity('Simple', 5760524373458944) {'u': <Entity {'auth_domain': 'gmail.com', 'federated_identity': 'A', 'email': 'daniel.j.hermes@gmail.com', 'federated_provider': 'B', 'user_id': '...redacted...'}>}>
```

**Comment by dhermes** [comment]
@chrisrossi PTAL. I have tentatively removed the "do not merge" label. I think this is good to go but we should have a Googler (e.g. @theacodes or @crwilcox) weigh in on the newly added `User` class.

File: ndb/src/google/cloud/ndb/model.py [hunk 1]
```diff
@@ -2352,10 +2361,317 @@ def _from_base_type(self, value):
...
+
+        The ``auto_current_user`` and ``auto_current_user_add`` arguments are
+        no longer supported.
+
+    .. automethod:: _validate
+    .. automethod:: _prepare_for_put
+
+    Args:
+        name (str): The name of the property.
+        auto_current_user (bool): Indicates if the value should be indexed.
```
Comment (chrisrossi) [inline ndb/src/google/cloud/ndb/model.py]: Description seems off...

**Comment by chrisrossi** [comment]
LGTM with caveat that I don't have much expertise with how users are implemented in Google Cloud, so I'm assuming the bits that integrate with the larger cloud framework actually make sense.

File: ndb/src/google/cloud/ndb/model.py (hunk 1 above)
Comment (dhermes) [reply ndb/src/google/cloud/ndb/model.py]: Copy-pasta!

**Pull request merged**